backup/backup.go

```diff
@@ -0,0 +1,113 @@
+package backup
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/featureform/provider"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"io/ioutil"
+	"time"
+)
+
+// Client Allows ETCD to be tested with mock values
+type Client interface {
+	Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error)
+}
+
+type BackupManager struct {
+	ETCDClient Client
+	Provider   provider.FileStoreType
+	as         provider.Provider
+}
+
+func (b *BackupManager) Save(filename string) error {
+	err := b.takeSnapshot(filename)
+	if err != nil {
+		return fmt.Errorf("could not take snapshot: %v", err)
+	}
+
+	backupProvider, err := b.getBackupProvider()
+	if err != nil {
+		return err
+	}
+	err = backupProvider.Init()
+	if err != nil {
+		return fmt.Errorf("could not initialize provider: %v", err)
+	}
+	err = backupProvider.Upload(filename, filename)
+	if err != nil {
+		return fmt.Errorf("cannot upload snapshot to filestore: %v", err)
+	}
+	return nil
+}
+
+func (b *BackupManager) Restore() error {
+	panic(fmt.Errorf("Restore() unimplemented"))
+	return nil
+}
+
+func (b *BackupManager) getBackupProvider() (Provider, error) {
+	var backupProvider Provider
+	switch b.Provider {
+	case provider.Azure:
+		backupProvider = &Azure{}
```

Comment:
The user should set &Azure or &Local in the struct above. This allows them to overwrite any fields they want to.

Reply:
Can you clarify? Are you saying to remove the switch and just set the struct directly?

Reply:
Yes
BackupManager{client, &Azure{}}